File: String/Reverse_a_String.java
    // Program to Reverse a String in Java.
    // Input a String from the user. Reverse the String and print the output.
    /*
     * Algorithm:
     *
     * -Input a String from the user using the Scanner class.
     * -Declare and initialise a new String variable.
     * -Run a for loop that extracts each character of the inputted String in a backward manner,
     *  i.e. it first extracts the character at last index, then the one at 2nd last index, and so on.
     * -Keep adding the extracted characters in the new String.
     * -Once the for loop is complete, the new String will be the reverse of the inputted String. Print it.
     */

    import java.util.*;

    class Reverse_a_String
    {
        public static void main(String[] args) {
            
            // Inputting a String from the user
            Scanner sc= new Scanner(System.in);
            System.out.println("Enter a String");
            String str=sc.nextLine();
        
            // Declaring and Initialising a new String
            String rev=""; 
            
            int len=str.length();
        
            // for loop for extracting each character of the given string in a backward manner and adding it in the new string
            for(int i=len-1; i>=0; i--)
            {
              rev=rev+str.charAt(i);
            }
        
            // Printing the reversed String
            System.out.println(rev);
            sc.close();
        }
    }

    /*
     * INPUT 1 :
     * Enter a String: Reverse
     * 
     * OUTPUT 1 :
     * The Original String: Reverse
     * The Reversed String: esreveR
     * 
     * INPUT 2 :
     * Enter a String: Reverse a String
     * 
     * OUTPUT 2 :
     * The Original String: Reverse a String
     * The Reversed String: gnirtS a esreveR
    */
   
    /*
     * Time Complexity: O(N)
     * where N is the length of the given String
     * 
     * Space Complexity: O(N)
     * where N is the length of the given String
    */

File: String/Reverse_A_String.c
//C Program to reverse a string:
//Problem Statement: Given a string of size N. Reverse the string and display it.

/* Algorithm- 
To reverse a string :
1. Start
2. Declare a user-defined function revstr() 
3. Run a for loop using the swapping logic in definition of revstr()
4. Run another for loop for displaying the reversed string
5. Under main() take user input for string.
6. Call revstr() in main()
7. Stop
*/

#include<stdio.h> //included the standard input/output header file

//User-Defined Function Body containing logic:
void revstr(char s[]) /*declaring and defining user-defined function, revstr() 
                        which reverses a string by using swapping logic*/
{
    int a, n;         //initializing a counter variable 'a' and another 'n'
    int temp = 0;     //initializing a temporary variable 'temp' to store the swapped value
    for(n=0; s[n] != 0; n++); //a for loop to check if s[] is empty or not
    for(a=0; a<n/2; a++) //another for loop for non-empty string, where swapping takes place
    {
        temp = s[a];
        s[a]=s[n - 1 - a];
        s[n - 1 - a] = temp;
    }
    for(a = 0; a < n; a++) //another for loop, to display the reversed string
    
        putchar(s[a]);
        printf(" is the reversed string .");
    printf("\n");
}


//Main Function:
int main(void)
{
    char s[1000]; 
    printf("Insert the string to be reversed: ");
    scanf("%s", s); //Taking input string from user
    revstr(s); //Calling user-defined function revstr() from main()
    return 0;
}

/*
Test Case 1:
Input: 
Insert the string to be reversed: abc
Output:
cba is the reversed string .

Test Case 2:
Input: 
Insert the string to be reversed: arushi
Output: 
ihsrua is the reversed string .

Time Complexity: O(N) 
Space Complexity: O(1)
*/


